### PR TITLE
Add `Infix To Prefix` new algorithm with unit tests

### PR DESCRIPTION
**Pull request by Hardvan**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * Convert an infix expression to a prefix expression.
+     *
+     * @param infixExpression the infix expression to convert
+     * @return the prefix expression
+     * @throws Exception if the infix expression has unbalanced brackets
+     */
+    public static String infix2Prefix(String infixExpression) throws Exception {
+        if (!BalancedBrackets.isBalanced(filterBrackets(infixExpression))) {
+            throw new Exception("invalid expression");
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:20]: Will be better to use [IllegalArgumentException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalArgumentException.html)

File: src/test/java/com/thealgorithms/stacks/InfixToPrefixTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class InfixToPrefixTest {
+
+    @ParameterizedTest
+    @MethodSource("provideValidExpressions")
+    void testValidExpressions(String infix, String expectedPrefix) throws Exception {
+        assertEquals(expectedPrefix, InfixToPrefix.infix2Prefix(infix));
+    }
+
+    private static Stream<Arguments> provideValidExpressions() {
+        return Stream.of(Arguments.of("3+2", "+32"), // Simple addition
```
Comment (alxkm) [inline src/test/java/com/thealgorithms/stacks/InfixToPrefixTest.java:19]: It might be nice to also have tests for empty string and null value.

**Review by alxkm — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java (hunk 1 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:20]: Oh okay. Change implemented.

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Convert an infix expression to a prefix expression.
+     *
+     * @param infixExpression the infix expression to convert
+     * @return the prefix expression
+     * @throws IllegalArgumentException if the infix expression has unbalanced brackets
+     * @throws NullPointerException     if the infix expression is null
+     */
+    public static String infix2Prefix(String infixExpression) throws IllegalArgumentException {
+        // Check for unbalanced brackets
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:20]: This comment is redundant, since the method already provides such information.

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if (!BalancedBrackets.isBalanced(filterBrackets(infixExpression))) {
+            throw new IllegalArgumentException("invalid expression");
+        }
+        // Null input
+        if (infixExpression == null) {
+            throw new NullPointerException("null input");
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:26]: The same here: "Input expression cannot be null."

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+     * @return the prefix expression
+     * @throws IllegalArgumentException if the infix expression has unbalanced brackets
+     * @throws NullPointerException     if the infix expression is null
+     */
+    public static String infix2Prefix(String infixExpression) throws IllegalArgumentException {
+        // Check for unbalanced brackets
+        if (!BalancedBrackets.isBalanced(filterBrackets(infixExpression))) {
+            throw new IllegalArgumentException("invalid expression");
+        }
+        // Null input
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:24]: Superfluous comment.

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        if (infixExpression == null) {
+            throw new NullPointerException("null input");
+        }
+        // Empty string
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:28]: And here also. Superfluous comment.

File: src/main/java/com/thealgorithms/stacks/InfixToPrefix.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        if (infixExpression.isEmpty()) {
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/stacks/InfixToPrefix.java:29]: And it would be better to carry out the following sequence of checks:

```
    if (infixExpression == null) {
        throw new NullPointerException("Input expression cannot be null.");
    }

    if (infixExpression.trim().isEmpty()) {
        return "";
    }

    if (!BalancedBrackets.isBalanced(filterBrackets(infixExpression))) {
        throw new IllegalArgumentException("Invalid expression: unbalanced brackets.");
    }
```

**Review by alxkm — CHANGES_REQUESTED** [review]

**Comment by Hardvan** [comment]
I have implemented the suggested changes @alxkm .

**Review by alxkm — DISMISSED** [review]

**Review by alxkm — APPROVED** [review]
